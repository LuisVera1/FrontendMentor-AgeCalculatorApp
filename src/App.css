:root {
	/* primary colors */
	--purple: #854dff;
	--light-red: hsl(0, 100%, 67%);

	/* neutral colors */
	--off-white: #f0f0f0;
	--white: #ffffff;
	--smokey-gray: #716f6f;
	--light-gray: #dbdbdb;
	--off-black: #141414;
}

/* 
- Font size (inputs): 32px

- Family: [Poppins](https://fonts.google.com/specimen/Poppins)
- Weights: 400i, 700, 800i
*/

* {
	box-sizing: border-box;
	font-family: 'Poppins', sans-serif;
	margin: 0;
	padding: 0;
}

main {
	align-items: center;
	background-color: var(--off-white);
	display: flex;
	justify-content: center;
	min-height: 100vh;
}

/* --------------- card */
.card {
	/* height: 651px; */
	background-color: var(--white);
	border-radius: 24px 24px 200px;
	padding: 56px;
	width: 840px;
}

/* --------------- form */
.form {
	display: flex;
	flex-direction: row;
	gap: 32px;
}

.form__section {
	display: flex;
	flex-direction: column;
	gap: 8px;
}

.form__label {
	color: var(--smokey-gray);
	font-size: 0.875rem;
	font-weight: 700;
}

.form__label--error {
	color: var(--light-red);
	font-size: 0.875rem;
}

.form__input {
	border: none;
	border: 1px solid #15151540;
	border-radius: 8px;
	color: var(--off-black);
	font-size: 2rem;
	height: 72px;
	padding: 12px 24px;
	width: 160px;
}

.form__input::placeholder {
	color: #15151580;
}

.form__input:focus {
	border: 1px solid var(--purple);
	outline: none;
}

.form__input--error {
	border: 1px solid var(--light-red);
}

/* Removing arrows on inputs */

/* Chrome, Safari, Edge, Opera */
input::-webkit-outer-spin-button,
input::-webkit-inner-spin-button {
	-webkit-appearance: none;
	margin: 0;
}

/* Firefox */
input[type='number'] {
	appearance: textfield;
}

.form__input--error-msg {
	color: var(--light-red);
	font-size: 0.875rem;
	font-style: italic;
	font-weight: 400;
}

/* --------------- division */
.division {
	background: url('/line.webp');
	background-position: center;
	background-repeat: repeat-x;
}

.division__button {
	background-color: var(--purple);
	border: none;
	border-radius: 50%;
	/* margin-right: auto; */
	display: block;
	height: 96px;
	margin-left: auto;
	width: 96px;
}

.division__button:hover {
	background-color: var(--off-black);
	cursor: pointer;
}

.division-icon {
	color: var(--purple);
}

/* --------------- results */
.results-section {
	display: flex;
	gap: 8px;
}

.results-section__result {
	color: var(--purple);
	font-size: 6.5rem;
	font-style: italic;
	font-weight: 800;
	line-height: 110px;
}

.results-section__label {
	color: var(--off-black);
	font-size: 6.5rem;
	font-style: italic;
	font-weight: 800;
	line-height: 110px;
}
