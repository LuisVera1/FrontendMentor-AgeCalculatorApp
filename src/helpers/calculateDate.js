export const getYears = (sTime, eTime) => {
	const sDate = new Date(sTime.year, sTime.month, sTime.day);
	const eDate = new Date(eTime.year, eTime.month, eTime.day);
	const days = (eDate - sDate) / (1000 * 60 * 60 * 24);

	if (days < 0) return null;

	const years = Math.floor(days / 365);

	const exactDate = new Date(eTime.year - years, eTime.month, eTime.day);
	const remainDays = days - (eDate - exactDate) / (1000 * 60 * 60 * 24);

	return {
		years,
		remainDays,
		currentMonth: eTime.month,
		currentYear: eTime.year - years,
	};
};

const isLeapYear = (year) => {
	if (year % 4 == 0) {
		if (year % 100 == 0) {
			if (year % 400 == 0) {
				return true;
			}
		} else {
			return true;
		}
	} else {
		return false;
	}

	return false;
};

export const getMonths = ({ remainDays, currentMonth, currentYear }) => {
	const daysInMonths = {
		0: 31,
		1: isLeapYear(currentYear) ? 29 : 28,
		2: 31,
		3: 30,
		4: 31,
		5: 30,
		6: 31,
		7: 31,
		8: 30,
		9: 31,
		10: 30,
		11: 31,
	};

	let days = remainDays;
	let months = 0;
	let previusMonth = currentMonth == 0 ? 11 : currentMonth - 1;

	while (days > daysInMonths[previusMonth]) {
		months++;
		days -= daysInMonths[previusMonth];

		if (previusMonth == 0) {
			previusMonth = 11;
		} else {
			previusMonth -= 1;
		}
	}

	return { months, days };
};

/*   main function   */
export const calculateDate = (sTime, eTime) => {
	const yearsAndDays = getYears(sTime, eTime);
	if (!yearsAndDays) return null;

	const { years, remainDays, currentMonth, currentYear } = yearsAndDays;
	const monthsAndDays = getMonths({ remainDays, currentMonth, currentYear });
	const { months, days } = monthsAndDays;

	return { years, months, days };
};
