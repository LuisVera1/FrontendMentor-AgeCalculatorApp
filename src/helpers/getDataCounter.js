const getLastDigit = (number) => {
	const numStr = number.toString();
	const numberLength = numStr.length;
	return Number(numStr[numberLength - 1]);
};

export const getDataCounter = ({ years, months, days }) => {
	return {
		years: years,
		yearsI: years - getLastDigit(years),
		months: months,
		monthsI: months - getLastDigit(months),
		days: days,
		daysI: days - getLastDigit(days),
	};
};
