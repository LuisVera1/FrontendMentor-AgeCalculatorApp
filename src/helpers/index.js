export * from './calculateDate';
export * from './getDataCounter';
export * from './validateForm';
