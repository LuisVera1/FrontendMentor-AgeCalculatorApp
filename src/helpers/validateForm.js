export const errorMessages = {
	notValidDate: 'Must be a valid day',
	notValidMonth: 'Must be a valid month',
	notValidYear: 'Must be in the past',
};

/**
 * Validate inputs to eval if the value is logic
 * @param {*} value
 * @param {*} name
 * @example
 * year <= current year
 * months between 1 and 12
 * days between 1 and 31
 */
export const validateForm = ({ value, name }) => {
	let error = false;
	let message = '';

	// validate day
	if (name === 'day') {
		if (value >= 1 && value <= 31) {
			error = false;
			message = '';
		} else {
			error = true;
			message = errorMessages.notValidDate;
		}
	}

	// validate month
	if (name === 'month') {
		if (value >= 1 && value <= 12) {
			error = false;
			message = '';
		} else {
			error = true;
			message = errorMessages.notValidMonth;
		}
	}

	// validate year
	if (name === 'year') {
		const date = new Date();
		const year = date.getFullYear();

		if (value <= year) {
			error = false;
			message = '';
		} else {
			error = true;
			message = errorMessages.notValidYear;
		}
	}

	return { error, message };
};
