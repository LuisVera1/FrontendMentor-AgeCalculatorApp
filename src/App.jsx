import { useEffect, useState } from 'react'
import { calculateDate, getDataCounter, validateForm } from './helpers';
import './App.css'

const animationTemp = 70;
const steps = {
  start: 'start',
  calc: 'Cal',
  counter: 'counter',
}

const initialState = {
  day: {
    value: '',
    error: false,
    msg: ''
  },
  month: {
    value: '',
    error: false,
    msg: ''
  },
  year: {
    value: '',
    error: false,
    msg: ''
  },
  globalError: false,
  step: steps.start,
}

const initialResults = {
  years: '--',
  yearsI: '--',
  months: '--',
  monthsI: '--',
  days: '--',
  daysI: '--',
}

function App() {
  const [state, setState] = useState(initialState);
  const [results, setResults] = useState(initialResults);

  //Calculate differente
  useEffect(() => {
    if (state.step == steps.calc) {
      const { year, month, day } = state;
      const today = new Date();
      const sDate = new Date(year.value, month.value - 1, day.value);

      const result = calculateDate(sDate, today);
      if (result == null) {
        setState({ ...state, step: steps.start })
      } else {
        const valuesForCounter = getDataCounter(result)

        setState({ ...state, step: steps.counter })
        setResults(valuesForCounter)
      }
    }
  }, [state.step])

  //contols increments
  useEffect(() => {

    if (state.step == steps.counter) {
      setTimeout(() => {
        let resultsIncrement = { ...results }
        let difValues = 0;

        //days
        if (results.days > results.daysI) {
          resultsIncrement.daysI = resultsIncrement.daysI + 1;
          difValues++
        }

        //months
        if (results.months > results.monthsI) {
          resultsIncrement.monthsI = resultsIncrement.monthsI + 1;
          difValues++
        }

        //years
        if (results.years > results.yearsI) {
          resultsIncrement.yearsI = resultsIncrement.yearsI + 1;
          difValues++
        }

        if (difValues > 0) {
          setResults(resultsIncrement)
        } else {
          setState({ ...state, step: steps.start })
        }
      }, animationTemp)
    }
  }, [state.counter, results])

  const onInput = (event) => {
    const { value, name } = event.target;
    const { error, message } = validateForm({ value, name });

    const newState = {
      ...state,
      globalError: false,
      [name]: {
        value: value,
        error: error,
        msg: message
      }
    };
    setState(newState);
  }

  const handleSubmit = () => {
    if (state.step != steps.start) return;
    const { day, month, year } = state;

    let newState = { ...state }

    // when a field has an error
    if (state.day.error || state.month.error || state.year.error || state.globalError) return;

    // check empty fields
    let emptyFields = false;
    const fields = ['day', 'month', 'year']
    fields.forEach(field => {
      if (newState[field].value.length === 0) {
        newState[field].error = true;
        newState[field].msg = 'This field is required'
        emptyFields = true;
      }
    })

    //check invalid date (select more days per month)
    const inputDate = new Date(year.value, month.value - 1, day.value);

    //Check if the day did not move to the next month
    const intputDay = inputDate.getDate();
    let validDate = true;

    if (Number(day.value) !== intputDay && emptyFields == false) {
      newState.day.msg = 'Must be a valid date';
      newState.globalError = true;
      validDate = false;
    }

    //validate year less than 100
    const currentYear = new Date().getFullYear()
    if (year.value < 100 && year.value != '') {
      alert(`Only years between 100 - ${currentYear}`)
      return;
    }

    //calculate difference
    if (validDate == true && emptyFields == false) {
      newState.step = steps.calc;
    }

    setState(newState)
  }

  return (
    <>
      <main>

        <section className="card">

          {/* form */}
          <form className="form">
            {/* input-day */}
            <div className='form__section'>
              <label
                className={state.day.error || state.globalError ? 'form__label--error' : 'form__label'}
                htmlFor="day">
                DAY
              </label>
              <input
                className={state.day.error || state.globalError ? 'form__input form__input--error' : 'form__input'}
                data-testid="input-day"
                // inputProps={{ "data-testid": "input-day" }}
                name="day"
                type="number"
                placeholder="DD"
                value={state.day.value}
                onChange={onInput}
              />
              {(state.day.error || state.globalError) && <p data-testid="error-day" className='form__input--error-msg'>{state.day.msg}</p>}
            </div>

            {/* input-month */}
            <div className='form__section'>
              <label
                className={state.month.error || state.globalError ? 'form__label--error' : 'form__label'}
                htmlFor="month">
                MONTH
              </label>
              <input
                data-testid="input-month"
                className={state.month.error || state.globalError ? 'form__input form__input--error' : 'form__input'}
                name="month"
                type="number"
                placeholder="MM"
                value={state.month.value}
                onChange={onInput}
              />
              {state.month.error && <p data-testid="error-month" className='form__input--error-msg'>{state.month.msg}</p>}
            </div>

            {/* input-year */}
            <div className='form__section'>
              <label
                className={state.year.error || state.globalError ? 'form__label--error' : 'form__label'}
                htmlFor="year">
                YEAR
              </label>
              <input
                data-testid="input-year"
                className={state.year.error || state.globalError ? 'form__input form__input--error' : 'form__input'}
                name="year"
                type="number"
                placeholder="YYYY"
                value={state.year.value}
                onChange={onInput}
              />
              {state.year.error && <p data-testid="error-year" className='form__input--error-msg'>{state.year.msg}</p>}
            </div>
          </form>

          {/* button */}
          <div className="division">
            <button className='division__button'
              data-testid="button-submit"
              onClick={handleSubmit}>
              <img className='division-icon' height="46px" width="46px" src="/icon-arrow.svg" alt="calculate" />
            </button>
          </div>

          {/* results */}
          <div className='results'>

            <div className="results-section">
              <p data-testid="p-years" className='results-section__result'>{results.yearsI}</p>
              <p className='results-section__label'>years</p>
            </div>

            <div className="results-section">
              <p data-testid="p-months" className='results-section__result'>{results.monthsI}</p>
              <p className='results-section__label'>months</p>
            </div>

            <div className="results-section">
              <p data-testid="p-days" className='results-section__result'>{results.daysI}</p>
              <p className='results-section__label'>days</p>
            </div>

          </div>
        </section>
      </main>
    </>
  )
}

export default App
