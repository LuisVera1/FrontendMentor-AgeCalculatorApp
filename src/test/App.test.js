import {
	cleanup,
	fireEvent,
	getAllByTestId,
	getByTestId,
	render,
	screen,
} from '@testing-library/react';
import App from '../App';
import { errorMessages } from '../helpers';

describe('Test in <App />', () => {
	afterEach(cleanup);

	test('Must be match with the snapshot', () => {
		const { container } = render(<App />);

		expect(container).toMatchSnapshot();
	});

	test('Results must be displayed when a valid date is entered', () => {
		render(<App />);

		//empty values
		const pYears = screen.getByTestId('p-years').textContent;
		expect(pYears).toBe('--');

		const pMonths = screen.getByTestId('p-months').textContent;
		expect(pMonths).toBe('--');

		const pDays = screen.getByTestId('p-days').textContent;
		expect(pDays).toBe('--');

		//fill inputs
		fireEvent.change(screen.getByTestId('input-day'), {
			target: { value: '12' },
		});
		fireEvent.change(screen.getByTestId('input-month'), {
			target: { value: '2' },
		});
		fireEvent.change(screen.getByTestId('input-year'), {
			target: { value: '1998' },
		});

		//submit
		fireEvent.click(screen.getByTestId('button-submit'));

		//changed values
		const cpYears = screen.getByTestId('p-years').textContent;
		expect(cpYears).not.toBe('--');

		const cpMonths = screen.getByTestId('p-months').textContent;
		expect(cpMonths).not.toBe('--');

		const cpDays = screen.getByTestId('p-days').textContent;
		expect(cpDays).not.toBe('--');
	});

	test('Error must be displayed when submit a empty input', () => {
		render(<App />);

		fireEvent.click(screen.getByTestId('button-submit'));
		expect(screen.getAllByText('This field is required').length).toBe(3);
	});

	test('For Days, must accept only values between 1-31', () => {
		render(<App />);

		// input: 0
		fireEvent.change(screen.getByTestId('input-day'), {
			target: { value: 0 },
		});
		fireEvent.click(screen.getByTestId('button-submit'));

		expect(screen.getByTestId('error-day').textContent).toBe(
			errorMessages.notValidDate
		);

		// input: 32
		fireEvent.change(screen.getByTestId('input-day'), {
			target: { value: 32 },
		});
		fireEvent.click(screen.getByTestId('button-submit'));

		expect(screen.getByTestId('error-day').textContent).toBe(
			errorMessages.notValidDate
		);

		// input: 1 (No error)
		fireEvent.change(screen.getByTestId('input-day'), {
			target: { value: 1 },
		});
		fireEvent.click(screen.getByTestId('button-submit'));

		expect(screen.queryByTestId('error-day')).toBeNull();

		// input: 31 (No error)
		fireEvent.change(screen.getByTestId('input-day'), {
			target: { value: 31 },
		});
		fireEvent.click(screen.getByTestId('button-submit'));

		expect(screen.queryByTestId('error-day')).toBeNull();
	});

	test('For Months, must accept only values between 1-12', () => {
		render(<App />);

		// input: 0
		fireEvent.change(screen.getByTestId('input-month'), {
			target: { value: 0 },
		});
		fireEvent.click(screen.getByTestId('button-submit'));

		expect(screen.getByTestId('error-month').textContent).toBe(
			errorMessages.notValidMonth
		);

		// input: 13
		fireEvent.change(screen.getByTestId('input-month'), {
			target: { value: 13 },
		});
		fireEvent.click(screen.getByTestId('button-submit'));

		expect(screen.getByTestId('error-month').textContent).toBe(
			errorMessages.notValidMonth
		);

		// input: 1 (No error)
		fireEvent.change(screen.getByTestId('input-month'), {
			target: { value: 1 },
		});
		fireEvent.click(screen.getByTestId('button-submit'));

		expect(screen.queryByTestId('error-month')).toBeNull();

		// input: 12 (No error)
		fireEvent.change(screen.getByTestId('input-month'), {
			target: { value: 12 },
		});
		fireEvent.click(screen.getByTestId('button-submit'));

		expect(screen.queryByTestId('error-month')).toBeNull();
	});

	test('For years, must accept only values less than or equal to current year', () => {
		const date = new Date();
		const year = date.getFullYear();

		render(<App />);
		// input: current year -1 (No error)
		fireEvent.change(screen.getByTestId('input-year'), {
			target: { value: year - 1 },
		});
		fireEvent.click(screen.getByTestId('button-submit'));
		expect(screen.queryByTestId('error-year')).toBeNull();

		// input: current year (No error)
		fireEvent.change(screen.getByTestId('input-year'), {
			target: { value: year },
		});
		fireEvent.click(screen.getByTestId('button-submit'));
		expect(screen.queryByTestId('error-year')).toBeNull();

		// input: current year + 1
		fireEvent.change(screen.getByTestId('input-year'), {
			target: { value: year + 1 },
		});
		fireEvent.click(screen.getByTestId('button-submit'));

		expect(screen.getByTestId('error-year').textContent).toBe(
			errorMessages.notValidYear
		);
	});

	test('Must accept only valid dates (31-Abr -> X)', () => {
		render(<App />);

		fireEvent.change(screen.getByTestId('input-day'), {
			target: { value: 31 },
		});
		fireEvent.change(screen.getByTestId('input-month'), {
			target: { value: 4 },
		});
		fireEvent.change(screen.getByTestId('input-year'), {
			target: { value: 2000 },
		});

		fireEvent.click(screen.getByTestId('button-submit'));

		expect(screen.getByTestId('error-day')).toBeTruthy();
	});

	test('Must not accept furure dates', () => {
		const date = new Date();
		const year = date.getFullYear();
		const month = date.getMonth() + 1;
		const day = date.getDate() + 4;

		render(<App />);

		fireEvent.change(screen.getByTestId('input-day'), {
			target: { value: day },
		});
		fireEvent.change(screen.getByTestId('input-month'), {
			target: { value: month },
		});
		fireEvent.change(screen.getByTestId('input-year'), {
			target: { value: year },
		});

		fireEvent.click(screen.getByTestId('button-submit'));

		expect(screen.getByTestId('p-days').textContent).toBe('--');
		expect(screen.getByTestId('p-months').textContent).toBe('--');
		expect(screen.getByTestId('p-years').textContent).toBe('--');
	});
});
