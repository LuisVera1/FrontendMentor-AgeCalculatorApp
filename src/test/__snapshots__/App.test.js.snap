// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test in <App /> Must be match with the snapshot 1`] = `
<div>
  <main>
    <section
      class="card"
    >
      <form
        class="form"
      >
        <div
          class="form__section"
        >
          <label
            class="form__label"
            for="day"
          >
            DAY
          </label>
          <input
            class="form__input"
            data-testid="input-day"
            name="day"
            placeholder="DD"
            type="number"
            value=""
          />
        </div>
        <div
          class="form__section"
        >
          <label
            class="form__label"
            for="month"
          >
            MONTH
          </label>
          <input
            class="form__input"
            data-testid="input-month"
            name="month"
            placeholder="MM"
            type="number"
            value=""
          />
        </div>
        <div
          class="form__section"
        >
          <label
            class="form__label"
            for="year"
          >
            YEAR
          </label>
          <input
            class="form__input"
            data-testid="input-year"
            name="year"
            placeholder="YYYY"
            type="number"
            value=""
          />
        </div>
      </form>
      <div
        class="division"
      >
        <button
          class="division__button"
          data-testid="button-submit"
        >
          <img
            alt="calculate"
            class="division-icon"
            height="46px"
            src="/icon-arrow.svg"
            width="46px"
          />
        </button>
      </div>
      <div
        class="results"
      >
        <div
          class="results-section"
        >
          <p
            class="results-section__result"
            data-testid="p-years"
          >
            --
          </p>
          <p
            class="results-section__label"
          >
            years
          </p>
        </div>
        <div
          class="results-section"
        >
          <p
            class="results-section__result"
            data-testid="p-months"
          >
            --
          </p>
          <p
            class="results-section__label"
          >
            months
          </p>
        </div>
        <div
          class="results-section"
        >
          <p
            class="results-section__result"
            data-testid="p-days"
          >
            --
          </p>
          <p
            class="results-section__label"
          >
            days
          </p>
        </div>
      </div>
    </section>
  </main>
</div>
`;
